content/en-us/luau/type-checking.md

```diff
@@ -0,0 +1,326 @@
+---
+title: Luau Typing
+description: Luau uses gradual typing through the use
+of type annotations and inference.
+---
+
+Luau uses a typing engine which gives every variable an inferred type.
+
+```lua
+local foo = "bar" --> string
+local x = 5 --> number
+```
+
+## Defining a type
+A type can be defined using the `type` keyword
+
+```lua
+type Vector2 = {x: number, y: number}
+```
+
+## Inference Modes
+There are three type inference modes that Luau can use in `Class.Script|Scripts`
+
+* `--!nocheck` - Don't check types
+* `--!nonstrict` - Default mode for all scripts, only asserts variable types if they are explicitly annotated
+* `--!strict` - Asserts all types based off the inferred or explicitly annotated type
+
+## Types
+
+A type annoatation can be defined using the `:` operator after a variable, followed by a type definition, by default, in `nonstrict` mode, all variables are assigned the type `any`.
+
+```lua
+local foo: string = "bar"
+local x: number = 5
+```
+
+There are 4 primitive types that can be used in an annotation:
+* `nil` - no value
+* `boolean` - `true`/`false` value
+* `number` - a numeric value
+* `string` - text
+
+Functions and tables are defined differently, and within Roblox, all `Class.Instance|Instances` and `DataType.Enums` have their own types
+
+```lua
+part.Touched:Connect(function(otherPart: BasePart)
+	local brickColor: Enum.BrickColor = otherPart.BrickColor
+end)
+```
+
+An optional type can be defined using `?` at the end of the annotation.
+
+```lua
+local foo: string? = nil
+```
+
+## Function Typing
+
+Assume the following function
+```lua
+local function add(x, y)
+	return x + y
+end
+```
+
+This is a fairly simple function where the behavior is well defined. Take the value of `x` and add it to `y`. However, nothing is stopping someone from using a `string` type in the `add` function, which causes an error.
+
+This is because Luau doesn't know that we want to use numbers in this function, lets rework this function such that it does constrain `x` and `y` to be numbers.

```

Comment:
This should be attached to the sentence above either as part of the rephrase or as a separate thought.

```suggestion
This is because Luau doesn't know that we want to only use numbers in this function. Lets rework this function such that it does constrain `x` and `y` to be numbers.
```

Note that with the above change, the second sentence is too casual for documentation. You may want to rephrase this to state more directly what is to be done ("add number type annotations to x and y").